File: weird_captcha_gym/shared_runtime/server/incubator_graders/two_season_strand.py
from __future__ import annotations

import math
from typing import Any


MECHANIC_ID = "two_season_strand"


def _fail(feedback: str) -> dict[str, Any]:
    return {"graded": True, "passed": False, "feedback": feedback}


def _pairs(sequence: list[int], order: list[int]) -> list[list[int]]:
    stacks: dict[int, list[int]] = {0: [], 1: []}
    result: list[list[int]] = []
    for index in order:
        color = sequence[index]
        if color in (0, 1):
            stacks[color].append(index)
        else:
            opener = 1 if color == 2 else 0
            if stacks[opener]:
                result.append(sorted((stacks[opener].pop(), index)))
    return sorted(result)


def _integer_sequence(value: Any, length: int, label: str) -> list[int]:
    if not isinstance(value, list) or len(value) != length:
        raise ValueError(f"{label} has the wrong length")
    if any(isinstance(item, bool) or not isinstance(item, int) or item not in range(4) for item in value):
        raise ValueError(f"{label} contains an invalid bead colour")
    return list(value)


def _order(value: Any, length: int, label: str) -> list[int]:
    if not isinstance(value, list) or len(value) != length:
        raise ValueError(f"{label} order has the wrong length")
    if any(isinstance(item, bool) or not isinstance(item, int) for item in value):
        raise ValueError(f"{label} order contains a non-integer index")
    if sorted(value) != list(range(length)):
        raise ValueError(f"{label} order is not a permutation of the strand")
    return list(value)


def _pair_list(value: Any, length: int, label: str) -> list[list[int]]:
    if not isinstance(value, list):
        raise ValueError(f"{label} pair set is missing")
    pairs: list[list[int]] = []
    seen: set[int] = set()
    for raw in value:
        if not isinstance(raw, list) or len(raw) != 2:
            raise ValueError(f"{label} contains a malformed pair")
        left, right = raw
        if any(isinstance(item, bool) or not isinstance(item, int) for item in raw):
            raise ValueError(f"{label} contains a non-integer pair")
        if not 0 <= left < right < length:
            raise ValueError(f"{label} contains an out-of-range pair")
        if left in seen or right in seen:
            raise ValueError(f"{label} pairs one bead more than once")
        seen.update((left, right))
        pairs.append([left, right])
    if pairs != sorted(pairs):
        raise ValueError(f"{label} pair set is not canonical")
    return pairs


def _binding(payload: dict[str, Any], ground_truth: dict[str, Any], public_state: dict[str, Any]) -> str | None:
    for label, record in (("payload", payload), ("ground truth", ground_truth), ("public state", public_state)):
        if str(record.get("mechanic_id") or "") != MECHANIC_ID:
            return f"{label} mechanic mismatch"
    challenge_id = str(ground_truth.get("challenge_id") or "")
    task_id = str(ground_truth.get("task_id") or "")
    if not challenge_id or str(payload.get("challenge_id") or "") != challenge_id:
        return "stale challenge"
    if str(public_state.get("challenge_id") or "") != challenge_id:
        return "public challenge mismatch"
    if not task_id or str(payload.get("task_id") or "") != task_id:
        return "payload task mismatch"
    if str(public_state.get("task_id") or "") != task_id:
        return "public task mismatch"
    return None


def _contract(ground_truth: dict[str, Any], public_state: dict[str, Any]) -> dict[str, Any]:
    parameters = ground_truth.get("parameters")
    if not isinstance(parameters, dict) or parameters != public_state.get("parameters"):
        raise ValueError("public difficulty parameters differ from the hidden contract")
    length = int(parameters.get("strand_length") or 0)
    budget = int(parameters.get("edit_budget") or 0)
    if length % 8 or not 40 <= length <= 88 or not 2 <= budget <= 40:
        raise ValueError("strand dimensions are outside supported limits")
    initial = _integer_sequence(ground_truth.get("initial_sequence"), length, "initial strand")
    canonical = _integer_sequence(ground_truth.get("canonical_sequence"), length, "canonical strand")
    if public_state.get("initial_sequence") != initial:
        raise ValueError("public initial strand differs from the hidden contract")
    orders_raw = ground_truth.get("season_orders")
    if not isinstance(orders_raw, dict) or orders_raw != public_state.get("season_orders"):
        raise ValueError("season traversal contract mismatch")
    orders = {
        season: _order(orders_raw.get(season), length, season)
        for season in ("spring", "winter")
    }
    targets_raw = ground_truth.get("target_pairs")
    if not isinstance(targets_raw, dict) or targets_raw != public_state.get("target_pairs"):
        raise ValueError("public blueprints differ from the hidden contract")
    targets = {
        season: _pair_list(targets_raw.get(season), length, season)
        for season in ("spring", "winter")
    }
    for season in ("spring", "winter"):
        if _pairs(canonical, orders[season]) != targets[season]:
            raise ValueError(f"{season} blueprint is not generated by the canonical strand")
    return {
        "length": length,
        "budget": budget,
        "initial": initial,
        "orders": orders,
        "targets": targets,
    }


def _indices(value: Any, length: int) -> list[int]:
    if not isinstance(value, list) or not value:
        raise ValueError("edited bead indices are missing")
    if any(isinstance(item, bool) or not isinstance(item, int) or not 0 <= item < length for item in value):
        raise ValueError("edited bead index is invalid")
    if len(set(value)) != len(value):
        raise ValueError("edited bead indices repeat")
    return list(value)


def _pair_progress(sequence: list[int], orders: dict[str, list[int]], targets: dict[str, list[list[int]]]) -> dict[str, dict[str, int]]:
    result: dict[str, dict[str, int]] = {}
    for season in ("spring", "winter"):
        current = _pairs(sequence, orders[season])
        target_set = {tuple(pair) for pair in targets[season]}
        result[season] = {
            "paired": len(current),
            "matched": len(target_set & {tuple(pair) for pair in current}),
            "target": len(target_set),
        }
    return result


def grade(payload: dict[str, Any], ground_truth: dict[str, Any], public_state: dict[str, Any]) -> dict[str, Any]:
    binding_error = _binding(payload, ground_truth, public_state)
    if binding_error:
        return _fail(binding_error)
    truth_condition = ground_truth.get("control_condition")
    if truth_condition != public_state.get("control_condition"):
        return _fail("public control condition differs from the hidden contract")
    interaction = str((truth_condition or {}).get("interaction") or "full")
    if interaction not in {"simplified", "full"}:
        return _fail("interaction condition is invalid")
    if str(payload.get("interaction_mode") or "") != interaction:
        return _fail("payload interaction mode mismatch")
    try:
        contract = _contract(ground_truth, public_state)
    except (TypeError, ValueError) as exc:
        return _fail(f"invalid strand contract: {exc}")

    events = payload.get("edits")
    if not isinstance(events, list) or len(events) > 80:
        return _fail("edit transcript is missing or outside limits")
    sequence = contract["initial"][:]
    edit_count = 0
    for sequence_number, event in enumerate(events, start=1):
        if not isinstance(event, dict) or event.get("sequence") != sequence_number:
            return _fail(f"edit {sequence_number} has an invalid sequence number")
        source = str(event.get("input_source") or "")
        allowed = {"palette_apply"} if interaction == "simplified" else {"strand_drag"}
        if source not in allowed:
            return _fail(f"edit {sequence_number} uses the wrong interaction input")
        color = event.get("color")
        if isinstance(color, bool) or not isinstance(color, int) or color not in range(4):
            return _fail(f"edit {sequence_number} has an invalid colour")
        try:
            indices = _indices(event.get("indices"), contract["length"])
        except ValueError as exc:
            return _fail(f"edit {sequence_number} is malformed: {exc}")
        if source == "palette_apply" and len(indices) != 1:
            return _fail(f"edit {sequence_number} must affect exactly one bead")
        if source == "strand_drag":
            if any(abs(left - right) != 1 for left, right in zip(indices, indices[1:])):
                return _fail(f"edit {sequence_number} is not one contiguous paint stroke")
            gesture = event.get("gesture")
            if not isinstance(gesture, dict):
                return _fail(f"edit {sequence_number} is missing its drag evidence")
            samples = gesture.get("sample_count")
            travel = gesture.get("travel_px")
            if isinstance(samples, bool) or not isinstance(samples, int) or samples < max(3, len(indices)):
                return _fail(f"edit {sequence_number} has too few pointer samples")
            if isinstance(travel, bool) or not isinstance(travel, (int, float)) or not math.isfinite(float(travel)):
                return _fail(f"edit {sequence_number} has invalid pointer travel")
            if float(travel) < max(12.0, 10.0 * (len(indices) - 1)):
                return _fail(f"edit {sequence_number} drag is shorter than its painted run")
            if gesture.get("start_index") != indices[0] or gesture.get("end_index") != indices[-1]:
                return _fail(f"edit {sequence_number} drag endpoints do not match its painted run")
        changed = [index for index in indices if sequence[index] != color]
        if not changed:
            return _fail(f"edit {sequence_number} does not change the strand")
        for index in indices:
            sequence[index] = color
        edit_count += len(changed)
        if edit_count > contract["budget"]:
            return _fail(f"edit budget exceeded at edit {sequence_number}")
        if event.get("changed_count") != len(changed):
            return _fail(f"edit {sequence_number} has an inconsistent changed-bead count")
        expected_progress = _pair_progress(sequence, contract["orders"], contract["targets"])
        if event.get("pair_progress_after") != expected_progress:
            return _fail(f"edit {sequence_number} has inconsistent fold feedback")

    current_pairs = {
        season: _pairs(sequence, contract["orders"][season])
        for season in ("spring", "winter")
    }
    if payload.get("final_sequence") != sequence:
        return _fail("submitted strand does not match replay")
    if payload.get("folds") != current_pairs:
        return _fail("submitted seasonal folds do not match replay")
    if payload.get("edit_count") != edit_count:
        return _fail("submitted edit count does not match replay")
    exact = all(current_pairs[season] == contract["targets"][season] for season in ("spring", "winter"))
    passed = payload.get("completed") is True and exact and 0 < edit_count <= contract["budget"]
    progress = _pair_progress(sequence, contract["orders"], contract["targets"])
    return {
        "graded": True,
        "passed": passed,
        "feedback": (
            f"two-season replay: Spring {progress['spring']['matched']}/{progress['spring']['target']} stems; "
            f"Winter {progress['winter']['matched']}/{progress['winter']['target']} stems; "
            f"edits {edit_count}/{contract['budget']}"
        ),
    }


def cheat(public_state: dict[str, Any], ground_truth: dict[str, Any]) -> dict[str, Any]:
    return {
        "canonical_sequence": ground_truth.get("canonical_sequence") or [],
        "mutated_indices": ground_truth.get("mutated_indices") or [],
        "canonical_paint_run": ground_truth.get("canonical_paint_run") or [],
        "instruction": "Restore the canonical bead colours, then seal both seasonal folds.",
        "answers": [],
    }

File: weird_captcha_gym/shared_runtime/app/mechanics/two_season_strand.js
(() => {
  "use strict";

  const SEASONS = ["spring", "winter"];
  const model = {
    state: null,
    sequence: [],
    edits: [],
    editCount: 0,
    brush: 0,
    drag: null,
    activeSeason: "spring",
    interaction: "full",
    terminal: false,
    busy: false,
    ready: false,
    helpers: null,
  };

  const clean = (value) => String(value == null ? "" : value)
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;")
    .replaceAll('"', "&quot;");
  const pairKey = (pair) => `${pair[0]}:${pair[1]}`;
  const clonePairs = (pairs) => (pairs || []).map((pair) => [...pair]);

  function fold(sequence, order) {
    const stacks = {0: [], 1: []};
    const pairs = [];
    order.forEach((index) => {
      const color = sequence[index];
      if (color === 0 || color === 1) {
        stacks[color].push(index);
      } else {
        const opener = color === 2 ? 1 : 0;
        if (stacks[opener].length) {
          pairs.push([stacks[opener].pop(), index].sort((left, right) => left - right));
        }
      }
    });
    return pairs.sort((left, right) => left[0] - right[0] || left[1] - right[1]);
  }

  function currentFolds() {
    return Object.fromEntries(SEASONS.map((season) => [
      season,
      fold(model.sequence, model.state.season_orders[season]),
    ]));
  }

  function pairProgress() {
    const folds = currentFolds();
    return Object.fromEntries(SEASONS.map((season) => {
      const target = new Set(model.state.target_pairs[season].map(pairKey));
      const matched = folds[season].filter((pair) => target.has(pairKey(pair))).length;
      return [season, {paired: folds[season].length, matched, target: target.size}];
    }));
  }

  function isExact() {
    const folds = currentFolds();
    return SEASONS.every((season) => JSON.stringify(folds[season]) === JSON.stringify(model.state.target_pairs[season]));
  }

  function layout(season) {
    const order = model.state.season_orders[season];
    const result = {};
    order.forEach((index, rank) => {
      const angle = -Math.PI / 2 + (rank / order.length) * Math.PI * 2;
      if (season === "spring") {
        const radiusX = 158 + 24 * Math.sin(angle * 3);
        const radiusY = 108 + 9 * Math.cos(angle * 2);
        result[index] = {
          x: 260 + radiusX * Math.cos(angle),
          y: 143 + radiusY * Math.sin(angle),
        };
      } else {
        const radius = 112 + 24 * Math.cos(angle * 6);
        result[index] = {
          x: 260 + radius * Math.cos(angle),
          y: 143 + radius * Math.sin(angle),
        };
      }
    });
    return result;
  }

  function edgeMarkup(pair, positions, className) {
    const left = positions[pair[0]];
    const right = positions[pair[1]];
    const bend = Math.min(44, Math.hypot(right.x - left.x, right.y - left.y) * 0.14);
    const midpointX = (left.x + right.x) / 2;
    const midpointY = (left.y + right.y) / 2;
    const towardCenterX = (260 - midpointX) * 0.34;
    const towardCenterY = (143 - midpointY) * 0.34;
    const controlX = midpointX + towardCenterX + bend * 0.04;
    const controlY = midpointY + towardCenterY;
    return `<path class="${className}" d="M ${left.x.toFixed(2)} ${left.y.toFixed(2)} Q ${controlX.toFixed(2)} ${controlY.toFixed(2)} ${right.x.toFixed(2)} ${right.y.toFixed(2)}" />`;
  }

  function foldMarkup(season) {
    const positions = layout(season);
    const current = currentFolds()[season];
    const target = model.state.target_pairs[season];
    const currentKeys = new Set(current.map(pairKey));
    const targetKeys = new Set(target.map(pairKey));
    const mismatchNodes = new Set();
    current.filter((pair) => !targetKeys.has(pairKey(pair))).forEach((pair) => pair.forEach((index) => mismatchNodes.add(index)));
    target.filter((pair) => !currentKeys.has(pairKey(pair))).forEach((pair) => pair.forEach((index) => mismatchNodes.add(index)));
    const guidance = model.state.parameters.blueprint_guidance;
    const labelStride = Number(model.state.parameters.index_label_stride || 4);
    const order = model.state.season_orders[season];
    const backbone = order.map((index) => `${positions[index].x.toFixed(2)},${positions[index].y.toFixed(2)}`).join(" ");
    const blueprintEdges = target.map((pair) => edgeMarkup(pair, positions, currentKeys.has(pairKey(pair)) ? "fold-edge blueprint matched" : "fold-edge blueprint missing")).join("");
    const liveEdges = current.map((pair) => edgeMarkup(pair, positions, targetKeys.has(pairKey(pair)) ? "fold-edge live matched" : "fold-edge live extra")).join("");
    const nodes = order.map((index, rank) => {
      const point = positions[index];
      const color = model.state.palette[model.sequence[index]];
      const mismatch = mismatchNodes.has(index);
      const showLabel = index % labelStride === 0 || (mismatch && guidance !== "sparse") || (guidance === "strong" && rank % 2 === 0);
      return `<g class="fold-node${mismatch ? " is-mismatch" : ""}" transform="translate(${point.x.toFixed(2)} ${point.y.toFixed(2)})">
        <circle r="${mismatch ? 6.2 : 4.5}" fill="${clean(color.color)}" />
        ${showLabel ? `<text x="7" y="-5">${index + 1}</text>` : ""}
      </g>`;
    }).join("");
    const matchedCount = current.filter((pair) => targetKeys.has(pairKey(pair))).length;
    return `
      <article id="strand-panel-${season}" class="season-card season-${season}" data-season="${season}" role="tabpanel" aria-labelledby="strand-tab-${season}">
        <header><div><small>${season === "spring" ? "I · OUTWARD READING" : "II · BRAIDED READING"}</small><h2>${season.toUpperCase()}</h2></div><strong>${matchedCount}<i>/ ${target.length}</i></strong></header>
        <svg class="fold-map" viewBox="0 0 520 286" role="img" aria-label="${season} current fold over target blueprint">
          <polyline class="fold-backbone" points="${backbone}" />
          <g class="blueprint-layer">${blueprintEdges}</g>
          <g class="live-layer">${liveEdges}</g>
          <g class="node-layer">${nodes}</g>
        </svg>
        <footer><span><i class="legend-line target"></i>GHOST BLUEPRINT</span><span><i class="legend-line live"></i>CURRENT STEM</span><span><i class="legend-line wrong"></i>DISPLACED</span></footer>
      </article>`;
  }

  function seasonTabsMarkup() {
    const progress = pairProgress();
    return `<nav class="strand-season-tabs" role="tablist" aria-label="Season fold views">
      ${SEASONS.map((season, position) => {
        const active = model.activeSeason === season;
        const tally = progress[season];
        return `<button type="button" id="strand-tab-${season}" class="strand-season-tab season-${season}${active ? " is-active" : ""}" data-season-tab="${season}" role="tab" aria-selected="${active}" aria-controls="strand-panel-${season}" tabindex="${active ? "0" : "-1"}">
          <small>${position === 0 ? "I · OUTWARD" : "II · BRAIDED"}</small>
          <strong>${season.toUpperCase()}</strong>
          <span>${tally.matched}/${tally.target} STEMS</span>
        </button>`;
      }).join("")}
      <p>SWITCH VIEWS TO CHECK BOTH READINGS</p>
    </nav>`;
  }

  function beadMarkup(colorIndex, index) {
    const color = model.state.palette[colorIndex];
    return `<button type="button" class="strand-bead" data-index="${index}" data-color="${colorIndex}" style="--bead:${clean(color.color)}" aria-label="bead ${index + 1}, ${clean(color.name)}">
      <span>${clean(color.short)}</span><small>${index + 1}</small>
    </button>`;
  }

  function strandMarkup() {
    return model.sequence.map(beadMarkup).join("");
  }

  function paletteMarkup() {
    return model.state.palette.map((color) => `
      <button type="button" class="strand-swatch${model.brush === color.id ? " is-active" : ""}" data-palette-color="${color.id}" style="--swatch:${clean(color.color)}">
        <i></i><span>${clean(color.name)}</span><b>${clean(color.short)}</b>
      </button>`).join("");
  }

  function refresh() {
    const folds = document.getElementById("strand-folds");
    const strand = document.getElementById("strand-ribbon");
    const palette = document.getElementById("strand-palette");
    if (folds) folds.innerHTML = seasonTabsMarkup() + foldMarkup(model.activeSeason);
    if (strand) strand.innerHTML = strandMarkup();
    if (palette) palette.innerHTML = paletteMarkup();
    bindControls();
    const remaining = document.getElementById("strand-edits-left");
    if (remaining) remaining.textContent = String(Number(model.state.parameters.edit_budget) - model.editCount);
    model.ready = isExact();
    const seal = document.getElementById("strand-seal");
    seal?.classList.toggle("is-ready", model.ready);
    document.querySelector(".two-season-strand")?.classList.toggle("is-ready", model.ready);
  }

  function clearFreshFailure() {
    document.querySelector(".strand-verdict-fail")?.remove();
    document.querySelector(".two-season-strand")?.classList.remove("is-fresh-fail");
  }

  async function submit(completed) {
    if (model.busy || model.terminal) return;
    model.busy = true;
    document.querySelectorAll("button").forEach((button) => { button.disabled = true; });
    model.helpers.setReadout("CHECKING BOTH SEASONS…", "pending");
    const payload = {
      mechanic_id: model.state.mechanic_id,
      task_id: model.state.task_id,
      challenge_id: model.state.challenge_id,
      interaction_mode: model.interaction,
      edits: model.edits,
      events: model.edits,
      final_sequence: [...model.sequence],
      folds: currentFolds(),
      edit_count: model.editCount,
      completed,
    };
    try {
      const response = await fetch("/result", {
        method: "POST",
        headers: {"content-type": "application/json"},
        body: JSON.stringify(payload),
      });
      const outcome = await response.json();
      if (outcome.passed === true) {
        model.terminal = true;
        document.querySelector(".two-season-strand")?.classList.add("is-pass");
        document.querySelector(".two-season-strand")?.insertAdjacentHTML("beforeend", '<div class="strand-verdict strand-verdict-pass"><small>DUAL FOLD ACCEPTED</small><strong>PASS</strong></div>');
        model.helpers.setReadout("PASS · BOTH SEASONS HOLD", "passed");
      } else if (outcome.passed === false && outcome.state) {
        await model.helpers.render(outcome.state);
        const shell = document.querySelector(".two-season-strand");
        shell?.classList.add("is-fresh-fail");
        shell?.insertAdjacentHTML("beforeend", '<div class="strand-verdict strand-verdict-fail"><small>FOLD REJECTED · FRESH STRAND</small><strong>FAIL</strong></div>');
        model.helpers.setReadout("FAIL · FRESH STRAND ISSUED", "error");
      } else {
        model.busy = false;
        document.querySelectorAll("button").forEach((button) => { button.disabled = false; });
        model.helpers.setReadout("FAIL · FOLD SERVICE OFFLINE", "error");
      }
    } catch (_error) {
      model.busy = false;
      document.querySelectorAll("button").forEach((button) => { button.disabled = false; });
      model.helpers.setReadout("FAIL · FOLD SERVICE OFFLINE", "error");
    }
  }

  function applyColor(indices, color, inputSource, gesture = null) {
    if (model.busy || model.terminal) return;
    clearFreshFailure();
    const changed = indices.filter((index) => model.sequence[index] !== color);
    if (!changed.length) {
      model.helpers.setReadout("NO CHANGE · THAT COLOUR IS ALREADY SET", "idle");
      return;
    }
    const projected = model.editCount + changed.length;
    if (projected > Number(model.state.parameters.edit_budget)) {
      document.querySelector(".two-season-strand")?.insertAdjacentHTML("beforeend", '<div class="strand-verdict strand-verdict-fail strand-verdict-snap"><small>EDIT ALLOWANCE EXHAUSTED</small><strong>SNAP</strong></div>');
      model.helpers.setReadout("THREAD SNAP · EDIT ALLOWANCE EXCEEDED", "error");
      void submit(false);
      return;
    }
    indices.forEach((index) => { model.sequence[index] = color; });
    model.editCount = projected;
    const event = {
      sequence: model.edits.length + 1,
      indices: [...indices],
      color,
      changed_count: changed.length,
      input_source: inputSource,
      pair_progress_after: pairProgress(),
    };
    if (gesture) event.gesture = gesture;
    model.edits.push(event);
    refresh();
    document.querySelectorAll(".season-card").forEach((card) => {
      card.classList.remove("is-refolding");
      void card.offsetWidth;
      card.classList.add("is-refolding");
    });
    const progress = pairProgress();
    if (model.ready) {
      model.helpers.setReadout(`BOTH SEASONS HOLD · ${model.editCount} EDITS · SEAL IT`, "passed");
    } else {
      model.helpers.setReadout(`REFOLDED · SPRING ${progress.spring.matched}/${progress.spring.target} · WINTER ${progress.winter.matched}/${progress.winter.target}`, "idle");
    }
  }

  function paintSimplified(index) {
    if (model.interaction !== "simplified" || model.busy || model.terminal) return;
    applyColor([index], model.brush, "palette_apply");
  }

  function chooseSeason(season) {
    if (!SEASONS.includes(season) || season === model.activeSeason || model.busy || model.terminal) return;
    model.activeSeason = season;
    refresh();
    model.helpers.setReadout(`${season.toUpperCase()} VIEW · CHECK THE LIVE FOLD AGAINST ITS GHOST BLUEPRINT`, "idle");
  }

  function startDrag(event, node) {
    if (model.interaction !== "full" || event.button !== 0 || model.busy || model.terminal || model.drag) return;
    event.preventDefault();
    const index = Number(node.dataset.index);
    model.drag = {
      pointerId: event.pointerId,
      origin: index,
      visited: [index],
      startX: event.clientX,
      startY: event.clientY,
      lastX: event.clientX,
      lastY: event.clientY,
      travel: 0,
      samples: 1,
      node,
    };
    node.classList.add("is-dragging");
    node.style.setProperty("--stroke", model.state.palette[model.brush].color);
    model.helpers.setReadout(`BRUSH DOWN · ${model.state.palette[model.brush].name} · STROKE CONTIGUOUS BEADS`, "idle");
    try { node.setPointerCapture(event.pointerId); } catch (_error) { /* optional */ }
  }

  function moveDrag(event) {
    const drag = model.drag;
    if (!drag || drag.pointerId !== event.pointerId) return;
    event.preventDefault();
    drag.travel += Math.hypot(event.clientX - drag.lastX, event.clientY - drag.lastY);
    drag.lastX = event.clientX;
    drag.lastY = event.clientY;
    drag.samples += 1;
    const hit = document.elementsFromPoint(event.clientX, event.clientY).find((candidate) => (
      candidate.classList?.contains("strand-bead") && candidate !== drag.node
    ));
    if (!hit) return;
    const index = Number(hit.dataset.index);
    const last = drag.visited[drag.visited.length - 1];
    if (index !== last && Math.abs(index - last) === 1 && !drag.visited.includes(index)) {
      drag.visited.push(index);
      model.helpers.setReadout(`STROKE ${drag.visited.length} BEADS · RELEASE TO PAINT`, "idle");
    }
  }

  function endDrag(event, cancelled = false) {
    const drag = model.drag;
    if (!drag || drag.pointerId !== event.pointerId) return;
    event.preventDefault();
    model.drag = null;
    drag.node.classList.remove("is-dragging");
    drag.node.style.removeProperty("--stroke");
    try { drag.node.releasePointerCapture(event.pointerId); } catch (_error) { /* optional */ }
    if (cancelled) return;
    if (drag.travel < 12) {
      model.helpers.setReadout("NO PAINT · HOLD AND STROKE THE LOADED BRUSH", "idle");
      return;
    }
    applyColor(drag.visited, model.brush, "strand_drag", {
      sample_count: drag.samples,
      travel_px: Math.round(drag.travel * 100) / 100,
      start_index: drag.origin,
      end_index: drag.visited[drag.visited.length - 1],
    });
  }

  function choosePalette(color) {
    if (model.busy || model.terminal) return;
    clearFreshFailure();
    model.brush = color;
    refresh();
    model.helpers.setReadout(`BRUSH LOADED · ${model.state.palette[color].name}`, "idle");
  }

  function bindControls() {
    document.querySelectorAll(".strand-season-tab").forEach((node) => {
      node.addEventListener("click", () => chooseSeason(String(node.dataset.seasonTab)));
      node.addEventListener("keydown", (event) => {
        if (!['ArrowLeft', 'ArrowRight'].includes(event.key)) return;
        event.preventDefault();
        const next = model.activeSeason === "spring" ? "winter" : "spring";
        chooseSeason(next);
        document.getElementById(`strand-tab-${next}`)?.focus();
      });
    });
    document.querySelectorAll(".strand-bead").forEach((node) => {
      if (model.interaction === "simplified") node.addEventListener("click", () => paintSimplified(Number(node.dataset.index)));
      node.addEventListener("pointerdown", (event) => startDrag(event, node));
      node.addEventListener("pointermove", moveDrag);
      node.addEventListener("pointerup", (event) => endDrag(event));
      node.addEventListener("pointercancel", (event) => endDrag(event, true));
    });
    document.querySelectorAll(".strand-swatch").forEach((node) => {
      node.addEventListener("click", () => choosePalette(Number(node.dataset.paletteColor)));
    });
  }

  async function render(state, helpers) {
    document.body.dataset.mechanic = "two-season-strand";
    document.body.dataset.cheatMode = helpers.isCheatMode() ? "true" : "false";
    const interaction = state.control_condition?.interaction || "full";
    Object.assign(model, {
      state,
      sequence: [...state.initial_sequence],
      edits: [],
      editCount: 0,
      brush: 0,
      drag: null,
      activeSeason: "spring",
      interaction,
      terminal: false,
      busy: false,
      ready: false,
      helpers,
    });
    const columns = Math.ceil(state.initial_sequence.length / 2);
    helpers.app.innerHTML = `
      <section class="two-season-strand" data-interaction="${clean(interaction)}" data-guidance="${clean(state.parameters.blueprint_guidance)}" data-challenge-id="${clean(state.challenge_id)}" style="--strand-columns:${columns}">
        <header class="strand-head">
          <div class="strand-title"><small>DUAL-SEASON WEAVING OFFICE · FOLD DESK 07</small><h1>TWO-SEASON STRAND</h1><p>${clean(state.prompt)}</p></div>
          <div class="strand-budget"><small>EDITS LEFT</small><strong id="strand-edits-left">${Number(state.parameters.edit_budget)}</strong><span>${state.initial_sequence.length} BEADS</span></div>
        </header>
        <main>
          <section class="strand-folds" id="strand-folds"></section>
          <section class="strand-console">
            <div class="strand-rule"><b>ONE STRAND · TWO READINGS</b><span>SUN↔NIGHT · MOSS↔BERRY · LAST OPEN STEM CLOSES FIRST</span></div>
            <div class="strand-ribbon" id="strand-ribbon">${strandMarkup()}</div>
            <aside class="strand-palette" id="strand-palette">${paletteMarkup()}</aside>
            <div class="strand-mode-note">${interaction === "full" ? "CHOOSE A COLOUR · HOLD AND STROKE THE BRUSH ACROSS ONE OR MORE CONTIGUOUS BEADS" : "CHOOSE A COLOUR · CLICK EACH BEAD TO PAINT IT"}</div>
          </section>
        </main>
        <footer class="strand-foot">
          <div class="strand-readout readout" data-status="idle">REFOLD DESK READY · REPAIR BOTH BLUEPRINTS</div>
          <button type="button" id="strand-seal">${clean(state.submit_label || "SEAL BOTH SEASONS")}</button>
        </footer>
        ${helpers.cheatPanelTemplate()}
      </section>`;
    refresh();
    document.getElementById("strand-seal")?.addEventListener("click", () => submit(model.ready));
    helpers.installCheatPanel();
    window.twoSeasonStrandModel = model;
  }

  window.WeirdCaptchaMechanics = window.WeirdCaptchaMechanics || {};
  window.WeirdCaptchaMechanics.two_season_strand = {
    rootSelector: ".two-season-strand",
    render,
  };
})();
